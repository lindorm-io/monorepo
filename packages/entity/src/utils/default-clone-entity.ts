import { isFunction } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { MetaColumnDecorator } from "../types";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

const reset: Array<MetaColumnDecorator> = [
  "CreateDateColumn",
  "ExpiryDateColumn",
  "DeleteDateColumn",
  "UpdateDateColumn",
  "VersionColumn",
];

export const defaultCloneEntity = <E extends IEntity>(
  Entity: Constructor<E>,
  entity: E,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const clone = new Entity();

  for (const column of metadata.columns) {
    if (metadata.generated.some((g) => g.key === column.key)) continue;

    if (
      reset.includes(column.decorator) ||
      metadata.primaryKeys.includes(column.key) ||
      metadata.generated.some((g) => g.key === column.key) ||
      metadata.indexes.some((u) => u.unique && u.keys.every((k) => k.key === column.key))
    ) {
      (clone as any)[column.key] = undefined;
    } else {
      (clone as any)[column.key] = parseColumn(column, entity);
    }

    if ((clone as any)[column.key]) continue;

    (clone as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return clone;
};
