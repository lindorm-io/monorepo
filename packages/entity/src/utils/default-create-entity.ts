import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

export const defaultCreateEntity = <
  E extends IEntity,
  O extends DeepPartial<E> = DeepPartial<E>,
>(
  Entity: Constructor<E>,
  options: O | E = {} as O,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const entity = new Entity();

  for (const column of metadata.columns) {
    (entity as any)[column.key] = parseColumn(column, entity, options);

    if ((entity as any)[column.key]) continue;

    (entity as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return entity;
};
