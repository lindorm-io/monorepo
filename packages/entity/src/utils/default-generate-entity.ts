import { Constructor, DeepPartial } from "@lindorm/types";
import { randomBytes, randomInt, randomUUID } from "crypto";
import { IEntity } from "../interfaces";
import { EntityMetadata, MetaGenerated } from "../types";
import { globalEntityMetadata } from "./global";

const generator = (config: Omit<MetaGenerated, "target">): any => {
  switch (config.strategy) {
    case "date":
      return new Date();

    case "float":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return Math.random() * (config.max - config.min) + config.min;

    case "integer":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return randomInt(config.min, config.max);

    case "string":
      config.length = config.length ?? 32;
      return randomBytes(config.length).toString("base64url");

    case "uuid":
      return randomUUID();

    default:
      return null;
  }
};

const validator = (
  config: Omit<MetaGenerated, "target">,
  metadata: EntityMetadata,
): void => {
  const column = metadata.columns.find((item) => item.key === config.key);

  if (!column) {
    throw new Error(`Column not found for key: ${config.key}`);
  }

  if (!column.type) return;

  switch (config.strategy) {
    case "date":
      if (column.type === "date") break;
      throw new Error(`Invalid column type for date generation: ${column.type}`);

    case "float":
      if (column.type === "float") break;
      throw new Error(`Invalid column type for int generation: ${column.type}`);

    case "integer":
      if (column.type === "integer") break;
      throw new Error(`Invalid column type for int generation: ${column.type}`);

    case "string":
      if (column.type === "string") break;
      throw new Error(`Invalid column type for string generation: ${column.type}`);

    case "uuid":
      if (column.type === "uuid") break;
      throw new Error(`Invalid column type for uuid generation: ${column.type}`);

    default:
      break;
  }
};

export const defaultGenerateEntity = <E extends IEntity>(
  target: Constructor<E>,
  entity: DeepPartial<E>,
): E => {
  const metadata = globalEntityMetadata.get(target);

  for (const config of metadata.generated) {
    if (entity[config.key]) continue;

    validator(config, metadata);

    (entity as any)[config.key] = generator(config);
  }

  return entity as E;
};
