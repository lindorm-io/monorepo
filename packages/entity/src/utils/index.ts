export * from "./default-clone-entity";
export * from "./default-create-entity";
export * from "./default-generate-entity";
export * from "./default-update-entity";
export * from "./default-validate-entity";
export * from "./get-collection-name";
export * from "./get-increment-name";
export * from "./get-save-strategy";
export * from "./global";
export * from "./remove-readonly";
export * from "./verify-readonly";
