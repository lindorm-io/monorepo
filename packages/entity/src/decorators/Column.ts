import { isObject, isString } from "@lindorm/is";
import { ColumnDecoratorOptions, MetaColumnType } from "../types";
import { globalEntityMetadata } from "../utils";

export function Column(
  type?: MetaColumnType,
  options?: Omit<ColumnDecoratorOptions, "type">,
): PropertyDecorator;
export function Column(options?: ColumnDecoratorOptions): PropertyDecorator;
export function Column(arg1?: any, arg2?: any): PropertyDecorator {
  return function (target, key) {
    const type = isString<MetaColumnType>(arg1)
      ? arg1
      : isObject<ColumnDecoratorOptions>(arg1)
        ? arg1.type
        : null;

    const options: ColumnDecoratorOptions | undefined = isObject(arg1) ? arg1 : arg2;

    globalEntityMetadata.addColumn({
      target: target.constructor,
      key: key.toString(),
      decorator: "Column",
      enum: options?.enum ?? null,
      fallback: options?.fallback ?? null,
      max: options?.max ?? null,
      min: options?.min ?? null,
      nullable: options?.nullable ?? false,
      optional: options?.optional ?? false,
      readonly: options?.readonly ?? false,
      schema: options?.schema ?? null,
      type: type ?? null,
    });
  };
}
