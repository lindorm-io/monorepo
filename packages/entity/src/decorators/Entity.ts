import { EntityDecoratorOptions } from "../types";
import { globalEntityMetadata } from "../utils";

export function Entity(options: EntityDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    globalEntityMetadata.addEntity({
      target,
      decorator: "Entity",
      cache: options?.cache || null,
      database: options?.database || null,
      name: options?.name || null,
      namespace: options?.namespace || null,
    });
  };
}
