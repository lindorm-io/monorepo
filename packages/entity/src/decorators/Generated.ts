import { isObject, isString } from "@lindorm/is";
import { GeneratedDecoratorOptions, MetaGeneratedStrategy } from "../types";
import { globalEntityMetadata } from "../utils";

export function Generated(
  strategy: MetaGeneratedStrategy,
  options?: Omit<GeneratedDecoratorOptions, "strategy">,
): PropertyDecorator;
export function Generated(options?: GeneratedDecoratorOptions): PropertyDecorator;
export function Generated(arg1?: any, arg2?: any): PropertyDecorator {
  return function (target, key) {
    const options = isObject(arg1)
      ? (arg1 as GeneratedDecoratorOptions)
      : ((arg2 ?? {}) as GeneratedDecoratorOptions);
    const strategy = isString(arg1) ? (arg1 as MetaGeneratedStrategy) : options.strategy;
    if (!strategy) {
      throw new Error("Generated strategy is required");
    }
    globalEntityMetadata.addGenerated({
      target: target.constructor,
      key: key as string,
      length: options.length ?? null,
      max: options.max ?? null,
      min: options.min ?? null,
      strategy,
    });
  };
}
