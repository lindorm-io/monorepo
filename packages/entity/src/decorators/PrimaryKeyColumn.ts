import { randomUUID } from "crypto";
import { globalEntityMetadata } from "../utils";

export function PrimaryKeyColumn(): PropertyDecorator {
  return function (target, key) {
    globalEntityMetadata.addColumn({
      target: target.constructor,
      key: key.toString(),
      decorator: "PrimaryKeyColumn",
      enum: null,
      fallback: () => randomUUID(),
      max: null,
      min: null,
      nullable: false,
      optional: false,
      readonly: true,
      schema: null,
      type: "uuid",
    });
    globalEntityMetadata.addPrimaryKey({
      target: target.constructor,
      key: key.toString(),
    });
  };
}
