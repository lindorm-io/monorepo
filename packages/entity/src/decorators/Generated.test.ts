import { globalEntityMetadata } from "../utils";
import { Column } from "./Column";
import { Entity } from "./Entity";
import { Generated } from "./Generated";
import { PrimaryKeyColumn } from "./PrimaryKeyColumn";

describe("Generated Decorator", () => {
  test("should add metadata", () => {
    @Entity()
    class GeneratedDecoratorEntity {
      @PrimaryKeyColumn()
      id!: string;

      @Column()
      @Generated("increment")
      one!: string;

      @Column()
      @Generated({ strategy: "integer", max: 4, min: 4 })
      two!: string;

      @Column()
      @Generated({ strategy: "string", length: 32 })
      three!: number;
    }

    expect(globalEntityMetadata.get(GeneratedDecoratorEntity)).toMatchSnapshot();
  });
});
