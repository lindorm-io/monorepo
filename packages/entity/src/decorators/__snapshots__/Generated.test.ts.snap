// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated Decorator should add metadata 1`] = `
{
  "columns": [
    {
      "decorator": "PrimaryKeyColumn",
      "enum": null,
      "fallback": [Function],
      "key": "id",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "readonly": true,
      "schema": null,
      "type": "uuid",
    },
    {
      "decorator": "Column",
      "enum": null,
      "fallback": null,
      "key": "one",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "readonly": false,
      "schema": null,
      "type": null,
    },
    {
      "decorator": "Column",
      "enum": null,
      "fallback": null,
      "key": "two",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "readonly": false,
      "schema": null,
      "type": null,
    },
    {
      "decorator": "Column",
      "enum": null,
      "fallback": null,
      "key": "three",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "readonly": false,
      "schema": null,
      "type": null,
    },
  ],
  "entity": {
    "cache": null,
    "database": null,
    "decorator": "Entity",
    "name": null,
    "namespace": null,
  },
  "extras": [],
  "generated": [
    {
      "key": "one",
      "length": null,
      "max": null,
      "min": null,
      "strategy": "increment",
    },
    {
      "key": "two",
      "length": null,
      "max": 4,
      "min": 4,
      "strategy": "integer",
    },
    {
      "key": "three",
      "length": 32,
      "max": null,
      "min": null,
      "strategy": "string",
    },
  ],
  "hooks": [],
  "indexes": [],
  "primaryKeys": [
    "id",
  ],
  "primarySource": null,
  "relations": [],
  "schemas": [],
}
`;
